bittensor/_dataset/dataset_impl.py

```diff
@@ -542,137 +537,494 @@ def set_data_size(self, batch_size, block_size):
+        '''
+        Checks the sample cache, and builds it if it is empty
+
+        Returns:
+            self.sample_cache (List[str]): 
+                The sample cache.
+        '''
+        if len(self.sample_cache) == 0:
+            sample_cat_params_list = random.sample(self.sample_cat_params_list, self.cache_size)
+            cache_sample_tasks = [self.cat(**sample_cat_params) for sample_cat_params in sample_cat_params_list]
+            self.sample_cache = [sample_bytes for sample_bytes in self.async_run(asyncio.gather(*cache_sample_tasks))]
 
-    def dataloader(self, epoch_length = 100):
-        r""" Creates a torch dataloader out of a subclass of this class.
+
+        return self.sample_cache
+
+    def __getitem__(self, idx: Optional[int]= None, filler_token:str='FILLER_TEXT') -> Union[List[str], torch.tensor]:
+        '''
+        Sample from queue or lazy loading. 
+        This involves sampling large text files that are then cached, generating
+        multiple samples per text file. When a threshold number of samples are generated
+        from the blocks, a FIFO priority is used to churn old blocks for new blocks, to avoid
+        data staleness.
 
         Args:
-            epoch_length (int, optional): 
-                A dataloader for a subset of the dataset of epoch_length is returned.
+            idx (int):
+                Sample index of dataset
+            filler_token (str):
+                Filler token to pad raw text 
+            
+        Returns:
+            output (Union[str, torch.tensor])
+
+
+        '''
+        # Random sampel idx if None.
+        if idx == None:
+            idx = random.randint(0, self.__len__())
 
+        # only sample if the cache is less than the cache_size
+        self.sample_count += 1
+        self.batch_count =  self.sample_count // self.batch_size
+        if self.run_generator :
+            raw_text_bytes = self.data_queue.get()
+        else:
+            self.resolve_sample_cache()
+            raw_text_bytes = self.sample_cache.pop(idx % len(self.sample_cache))
+        
+        try:
+            raw_text = raw_text_bytes.decode()
+        except UnicodeDecodeError as e:
+            raw_text = str(raw_text_bytes[2:-1])
+        if self.no_tokenizer:
+            raw_text =raw_text.split()
+            output = raw_text[:self.sequence_length]
+            remainder = self.sequence_length - len(output)
+            if remainder > 0:
+                output = output + [self.pad_token]*remainder
+            output = ' '.join(output)
+        else:
+            
+            tokenized_text =  self.tokenizer(raw_text, padding=True)['input_ids']
+            output = torch.tensor(tokenized_text)[:self.sequence_length]           
+            # Append the remainder with 0 (TODO use stop token)
+            seqeunce_length_remainder =  self.sequence_length - output.shape[0]
+            if seqeunce_length_remainder>0:
```

Comment:
We should pad from the left-hand side as thats the tokenizers behaviour by default.